src/providers/pkcs11/mod.rs

```diff
@@ -217,11 +220,17 @@ impl Provider {
             .map_err(to_response_status)?;
 
         if self.user_pin.is_some() {
+            let mut pin = self.user_pin.as_ref().unwrap().expose_secret().clone();
+            if pin.starts_with(PIN_HEX_PREFIX) {
+                if let Ok(raw_pin) = hex::decode(pin.split_off(PIN_HEX_PREFIX.len())) {
+                    pin = String::from_utf8_lossy(&raw_pin.as_slice()).to_string();
+                }
+            } else if pin.starts_with(PIN_STRING_PREFIX) {
+                pin = pin.split_off(PIN_STRING_PREFIX.len()).into();
+            }
```

Comment:
Since you're taking the pin out of the secret wrapper, you'll have to zeroize (after last use) both `pin` and `raw_pin`.

Reply:
Thanks for catching this ...
so for the `pin`, it should be `zeroize` after the login call, right?

Reply:
Yes, once the login finishes, you can clear it up.